File: packages/lynx-updates/android/com/sigx/updates/UpdatesLifecyclePublisher.kt
package com.sigx.updates

import android.util.Log
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.react.bridge.JavaOnlyMap
import com.lynx.tasm.LynxView
import java.util.UUID

/**
 * Per-LynxView publisher: pumps [UpdatesEventBus] payloads into JS
 * (`__sigxUpdatesEvent`) and registers the view with [UpdateStore] so
 * `applyNow` has a reload target. Instantiated by the generated
 * `GeneratedLifecyclePublishers.attachAll(lynxView)`.
 */
class UpdatesLifecyclePublisher(private val lynxView: LynxView) {

    private var token: UUID? = null

    fun attach() {
        UpdateStore.attachView(lynxView)
        token = UpdatesEventBus.addListener { payload -> publish(payload) }
    }

    fun detach() {
        token?.let { UpdatesEventBus.removeListener(it) }
        token = null
    }

    private fun publish(payload: JavaOnlyMap) {
        try {
            val params = JavaOnlyArray()
            params.pushMap(payload)
            lynxView.sendGlobalEvent(UpdatesEventBus.CHANNEL, params)
        } catch (e: Throwable) {
            Log.w(TAG, "publish failed: ${e.message}")
        }
    }

    private companion object {
        const val TAG = "SigxUpdates"
    }
}

File: packages/lynx-updates/ios/UpdatesLifecyclePublisher.swift
import Foundation
import UIKit
import Lynx

/// Per-LynxView publisher: pumps `UpdatesEventBus` payloads into JS
/// (`__sigxUpdatesEvent`), registers the view with `UpdateStore` so
/// `applyNow` has a reload target, and converts background→active app
/// transitions into `foreground` events for `checkOn: ['foreground']`.
/// Instantiated by the generated `GeneratedLifecyclePublishers.attachAll(to:)`.
final class UpdatesLifecyclePublisher {

    private weak var lynxView: LynxView?
    private var token: UUID?
    private var observers: [NSObjectProtocol] = []
    private var sawBackground = false

    init(lynxView: LynxView) {
        self.lynxView = lynxView
        UpdateStore.shared.attachView(lynxView)
        token = UpdatesEventBus.shared.addListener { [weak self] payload in
            guard let view = self?.lynxView else { return }
            view.sendGlobalEvent(UpdatesEventBus.channel, withParams: [payload])
        }

        // Foreground detection: only resume-after-background counts — cold
        // start is covered by the JS 'launch' trigger.
        let center = NotificationCenter.default
        observers.append(center.addObserver(
            forName: UIApplication.didEnterBackgroundNotification, object: nil, queue: .main
        ) { [weak self] _ in
            self?.sawBackground = true
        })
        observers.append(center.addObserver(
            forName: UIApplication.didBecomeActiveNotification, object: nil, queue: .main
        ) { [weak self] _ in
            guard let self, self.sawBackground else { return }
            self.sawBackground = false
            UpdatesEventBus.shared.emitForeground()
        })
    }

    deinit {
        if let token { UpdatesEventBus.shared.removeListener(token) }
        for observer in observers {
            NotificationCenter.default.removeObserver(observer)
        }
    }
}
